service/worker/deployment/deployment_reachability.go

```diff
@@ -0,0 +1,195 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2024 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deployment
+
+import (
+	"context"
+	"fmt"
+	"github.com/temporalio/sqlparser"
+	deploymentpb "go.temporal.io/api/deployment/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/server/common/cache"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	visibility_manager "go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/common/searchattribute"
+	"go.temporal.io/server/common/worker_versioning"
+	"time"
+)
+
+const (
+	reachabilityCacheOpenWFsTTL   = time.Minute
+	reachabilityCacheClosedWFsTTL = 10 * time.Minute
+	reachabilityCacheMaxSize      = 10000
+)
+
+/*
+const (
+	// Reachability level is not specified.
+	DEPLOYMENT_REACHABILITY_UNSPECIFIED DeploymentReachability = 0
+	// The deployment is reachable by new and/or open workflows. The deployment cannot be
+	// decommissioned safely.
+	DEPLOYMENT_REACHABILITY_REACHABLE DeploymentReachability = 1
+	// The deployment is not reachable by new or open workflows, but might be still needed by
+	// Queries sent to closed workflows. The deployment can be decommissioned safely if user does
+	// not query closed workflows.
+	DEPLOYMENT_REACHABILITY_CLOSED_WORKFLOWS_ONLY DeploymentReachability = 2
+	// The deployment is not reachable by any workflow because all the workflows who needed this
+	// deployment went out of retention period. The deployment can be decommissioned safely.
+	DEPLOYMENT_REACHABILITY_UNREACHABLE DeploymentReachability = 3
+)
+*/
+
+func getDeploymentReachability(
+	ctx context.Context,
+	namespaceEntry *namespace.Namespace,
+	seriesName, buildID, currentBuildID string,
```

Comment:
`currentBuildID` can be `nil` for those series that do not have a current deployment right?